src/main/scala/com/gu/zuora6for6modifier/Subscription.scala

```diff
@@ -125,25 +105,42 @@ object Subscription {
   private def validForPostponing(subscription: Subscription): Either[Throwable, Subscription] =
     for {
       _ <- test(subscription, _.ratePlans.length == 2)("Wrong number of plans")
-      _ <- test(
-        subscription,
-        _.ratePlans
-          .find(_.ratePlanName.startsWith(productRatePlanNamePrefix6For6))
-          .exists(_.lastChangeType.isEmpty)
-      )("No original 6 for 6 plan")
-      _ <- test(
-        subscription,
-        _.ratePlans
-          .find(_.ratePlanName.startsWith(productRatePlanNamePrefixMain))
-          .exists(_.lastChangeType.isEmpty)
-      )("No original main plan")
-      _ <- test(
-        subscription,
-        _.ratePlans
-          .find(_.ratePlanName.startsWith(productRatePlanNamePrefix6For6))
-          .exists(_.ratePlanCharges.headOption.exists { charge =>
-            charge.effectiveStartDate == Config.keyDate
-          })
-      )("Doesn't include key date")
+      _ <- test(subscription, _.ratePlans.find(isIntroPlan).exists(_.lastChangeType.isEmpty))(
+        "No original 6 for 6 plan"
+      )
+      _ <- test(subscription, _.ratePlans.find(isMainPlan).exists(_.lastChangeType.isEmpty))("No original main plan")
+      _ <- test(subscription, _.ratePlans.find(isIntroPlan).exists(hasChargeStartingOnDate(Config.keyDate)))(
+        "Doesn't include key date"
+      )
+    } yield subscription
+
+  def extractDataForPostponingMainRatePlan(
+      subscriptionName: String,
+      json: String
+  ): Either[Throwable, SubscriptionData] = {
+    def plusWeek(s: String): String = LocalDate.parse(s).plusWeeks(1).toString
+    for {
+      sub <- decode[Subscription](json)
+      valid <- validForPostponingMainRatePlan(sub)
+      planMain <- valid.ratePlans.find(isMainPlan).toRight(new RuntimeException("Can't find main plan"))
+    } yield SubscriptionData(
+      subscriptionName,
+      productPlanId6For6 = "",
```

Comment:
Is it worth using Option?

Reply:
Did consider this but I thought it would make the code less readable at the Json request body template.
An alternative is to have a different case class for each template that just contains what's required for that template.

Reply:
Have now introduced different models for each Json template so there are no redundant fields.